querysrv.go

```diff
@@ -0,0 +1,314 @@
+// Copyright (C) 2014-2015 Jakob Borg and Contributors (see the CONTRIBUTORS file).
+
+package main
+
+import (
+	"database/sql"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"log"
+	"net"
+	"net/url"
+	"time"
+
+	"github.com/golang/groupcache/lru"
+	"github.com/juju/ratelimit"
+	"github.com/syncthing/protocol"
+	"github.com/syncthing/syncthing/internal/discover"
+)
+
+type querysrv struct {
+	addr    *net.UDPAddr
+	db      *sql.DB
+	prep    map[string]*sql.Stmt
+	limiter *lru.Cache
+}
+
+func (s *querysrv) Serve() {
+	s.limiter = lru.New(lruSize)
+
+	conn, err := net.ListenUDP("udp", s.addr)
+	if err != nil {
+		log.Println("Listen:", err)
+		return
+	}
+
+	// Attempt to set the read and write buffers to 2^24 bytes (16 MB) or as high as
+	// possible.
+	for i := 24; i >= 16; i-- {
+		if conn.SetReadBuffer(1<<uint(i)) == nil {
+			break
+		}
+	}
+	for i := 24; i >= 16; i-- {
+		if conn.SetWriteBuffer(1<<uint(i)) == nil {
+			break
+		}
+	}
+
+	var buf = make([]byte, 1024)
+	for {
+		buf = buf[:cap(buf)]
+		n, addr, err := conn.ReadFromUDP(buf)
+		if err != nil {
+			log.Println("Read:", err)
+			return
+		}
+
+		if s.limit(addr) {
+			// Rate limit in effect for source
+			continue
+		}
+
+		if n < 4 {
+			log.Printf("Received short packet (%d bytes)", n)
+			continue
+		}
+
+		buf = buf[:n]
+		magic := binary.BigEndian.Uint32(buf)
+
+		switch magic {
+		case discover.AnnouncementMagic:
+			err := s.handleAnnounceV2(addr, buf)
+			globalStats.Announce()
+			if err != nil {
+				log.Println("Announce:", err)
+				globalStats.Error()
+			}
+
+		case discover.QueryMagic:
+			err := s.handleQueryV2(conn, addr, buf)
+			globalStats.Query()
+			if err != nil {
+				log.Println("Query:", err)
+				globalStats.Error()
+			}
+
+		default:
+			globalStats.Error()
+		}
+	}
+}
+
+func (s *querysrv) Stop() {
+	panic("stop unimplemented")
+}
+
+func (s *querysrv) handleAnnounceV2(addr *net.UDPAddr, buf []byte) error {
```

Comment:
so these are now technically `...AnnounceV3`, but that whole versioning thing is from we handled both v1 and v2 discovery in the same binary. Maybe drop the suffix?